### ISSUE TITLE
Unexpected results from bdev_ocf_get_bdevs

### ISSUE THREAD

**Issue by mikeBashStuff** [spdk/spdk#1320]
This has been seen while working on moving the .ini config to json in the `ocf_persistent_metadata` test.

There are two app runs to see if OCF data persisted on the nvme device. First run uses following json:
```
{
  "subsystems": [
    {
      "subsystem": "bdev",
      "config": [
        {
          "params": {
            "trtype": "PCIe",
            "name": "Nvme0",
            "traddr": "0000:82:00.0"
          },
          "method": "bdev_nvme_attach_controller"
        },
        {
          "method": "bdev_split_create",
          "params": {
            "base_bdev": "Nvme0n1",
            "split_count": 7,
            "split_size_mb": 128
          }
        },
        {
          "method": "bdev_ocf_create",
          "params": {
            "name": "ocfWT",
            "mode": "wt",
            "cache_bdev_name": "Nvme0n1p0",
            "core_bdev_name": "Nvme0n1p1"
          }
        },
        {
          "method": "bdev_ocf_create",
          "params": {
            "name": "ocfPT",
            "mode": "pt",
            "cache_bdev_name": "Nvme0n1p2",
            "core_bdev_name": "Nvme0n1p3"
          }
        },
        {
          "method": "bdev_ocf_create",
          "params": {
            "name": "ocfWB0",
            "mode": "wb",
            "cache_bdev_name": "Nvme0n1p4",
            "core_bdev_name": "Nvme0n1p5"
          }
        },
        {
          "method": "bdev_ocf_create",
          "params": {
            "name": "ocfWB1",
            "mode": "wb",
            "cache_bdev_name": "Nvme0n1p4",
            "core_bdev_name": "Nvme0n1p6"
          }
        }
      ]
    }
  ]
}
``` 
the `iscsi_tgt` loads this config without any hiccups and call to the `bdev_ocf_get_bdevs` returns the following:
```
[
  {
    "name": "ocfWT",
    "started": true,
    "cache": {
      "name": "Nvme0n1p0",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p1",
      "attached": true
    }
  },
  {
    "name": "ocfPT",
    "started": true,
    "cache": {
      "name": "Nvme0n1p2",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p3",
      "attached": true
    }
  },
  {
    "name": "ocfWB0",
    "started": true,
    "cache": {
      "name": "Nvme0n1p4",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p5",
      "attached": true
    }
  },
  {
    "name": "ocfWB1",
    "started": true,
    "cache": {
      "name": "Nvme0n1p4",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p6",
      "attached": true
    }
  }
]
```

Now, `iscsi_tgt` is killed and the started again with the following config:
```
{
  "subsystems": [
    {
      "subsystem": "bdev",
      "config": [
        {
          "params": {
            "trtype": "PCIe",
            "name": "Nvme0",
            "traddr": "0000:82:00.0"
          },
          "method": "bdev_nvme_attach_controller"
        },
        {
          "method": "bdev_split_create",
          "params": {
            "base_bdev": "Nvme0n1",
            "split_count": 7,
            "split_size_mb": 128
          }
        }
      ]
    }
  ]
}
```
The ocf creation is skipped since if `iscsi_tgt` was about to load the very same config as at the first run it would hang causing `waitforlisten()` to timeout (I think I will open another issue for that, because to me the app should just complain that all the ocf bdevs already exist instead of going unresponsive). 

The next call to `bdev_ocf_get_bdevs` results in the following:
```
[
  {
    "name": "ocfWT",
    "started": false,
    "cache": {
      "name": "Nvme0n1", # <- wut?
      "attached": false
    },
    "core": {
      "name": "Nvme0n1p1",
      "attached": true
    }
  },
  {
    "name": "ocfPT",
    "started": true,
    "cache": {
      "name": "Nvme0n1p2",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p3",
      "attached": true
    }
  },
  {
    "name": "ocfWB0",
    "started": true,
    "cache": {
      "name": "Nvme0n1p4",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p5",
      "attached": true
    }
  },
  {
    "name": "ocfWB1",
    "started": true,
    "cache": {
      "name": "Nvme0n1p4",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p6",
      "attached": true
    }
  }
]
```
The `iscsi_tgt` complains with the following:
```
bdev.c:5231:bdev_open: *ERROR*: Could not open Nvme0n1 - split module already claimed it
vbdev_ocf.c:1374:attach_base: *ERROR*: Unable to open device 'Nvme0n1' for writing
vbdev_ocf.c:1181:init_vbdev: *ERROR*: Device with name 'ocfWT' already exists
```

`vbdev_ocf.c:1181:init_vbdev: *ERROR*: Device with name 'ocfWT' already exists` - yes, all the ocf bdevs should already exist, but why complain only about this one? So `ocfWT` is not started, but it's cache device now suddenly points at `Nvme0n1` instead of `Nmve0n1p0` - why?  :D 

Now, since the initial complaint is about `split` module claiming the device, I tried to do the very same, but remove the `bdev_split_create` call from the second run, leaving only the call to attach to the controller. The `iscsi_tgt` doesn't complain about anything, but `bdev_ocf_get_bdevs` returns only this:
```
[
  {
    "name": "ocfWT",
    "started": false,
    "cache": {
      "name": "Nvme0n1",
      "attached": true
    },
    "core": {
      "name": "Nvme0n1p1",
      "attached": false
    }
  }
]
```
Which isn't really what we are looking for.

In short, both calls to `bdev_ocf_get_bdevs` should return the same results.

I don't really know what's wrong with the configs, I tried to move them 1:1 from the .init part, but apparently this is not correct. Any hints on what's going on would be appreciated. :)

**Comment by Donaim** [comment]
This looks to me like the metadata is getting in the way of creating OCF devices with RPC.
I will test if clearing the NVMe will help. (we already do that in examine tests)

**Comment by Donaim** [comment]
Apparently, Nvme0n1 metadata is the same as of Nvme0n1p0, so OCF creates one ocfWT based on Nvme0n1, (but it is claimed, so that is not possible) and then the second one based on Nvme0n1p0.

I think we should store cache name in OCF metadata as well, because currently only cores names are checked.

I will try to submit a fix today

**Comment by mikeBashStuff** [comment]
@Donaim Awesome, thanks for looking into this! :)

**Comment by Donaim** [comment]
I submitted a patch that fixes attaching to wrong cache device:  https://review.spdk.io/gerrit/c/spdk/spdk/+/1544
I will reproduce original issue to check if there is something else to this.

**Comment by mikeBashStuff** [comment]
@Donaim I just tested your patch with my changes and everything worked fine - second call to `bdev_ocf_get_bdevs` returned same data as the first one as expected. :)

**Comment by Donaim** [comment]
Fix merged

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_10398.html

**Comment by mikeBashStuff** [comment]
@Donaim I am afraid that there's still an issue - a bit different, but still in the realm of OCF. In here, https://ci.spdk.io/results/autotest-per-patch/builds/10567/archive/nvme-autotest/build.log you can see that the `ocf_persistent_metadata` passes without any issues, but here https://ci.spdk.io/results/autotest-per-patch/builds/10567/archive/nvme-phy-autotest/build.log it fails again while diffing the output from `bdev_ocf_get_bdevs`. This time, however, after a second run, all the bdevs are not started. I am not sure if this relates somehow to https://github.com/spdk/spdk/issues/1346 which is also seen during that test. Your input would be appreciated. :)

**Comment by mikeBashStuff** [comment]
Doesn't seem to be related to #1346 since this https://ci.spdk.io/public_build/autotest-per-patch_10591.html finished with a success. As opposed to this: https://ci.spdk.io/results/autotest-per-patch/builds/10591/archive/nvme-autotest/build.log where `vfio-pci` was not used.
